Add converter to convert v1 model checkpoint to v2 and an example notebook to use converter

hwpang opened the pull request:
## Description
@am2145 and I add a `v1_to_v2.py` file containing the utility functions to map a v1 checkpoint dictionary to a v2 checkpoint dictionary format. It converts the information from v1 arguments and v1 state dict to v2 hyperparameters and v2 state dict appropriately. As suggested by @KnathanM, we also add a jupyter notebook where the user can examine the contents of the checkpoint dictionary, convert the model, and load the converted model.

We test the converter using a regression model trained in v1. We compare the prediction made by the v1 model using the v1 code to the prediction made by the converted model using the v2 code, and the predictions are the same down to the 6th decimal place.

davidegraff commented:
great work and thanks for taking care of this!
- let's not put example notebooks in the core module. We don't want to distribute these in the source wheel, so we should move these to the top level `CHEMPROP_DIR/examples`
- I think our users would be well-served by making this into a CLI utility. I'm thinking somethin like so:
```
$ chemprop convert --help
usage: chemprop convert input_path [output_path]

positional arguments:
  input_path     the path of a saved v1 model

optional arguments:
  output_path  the path to which the converted model should be saved. If not provided, appends '_v2' to `input_path` and saves the new model to that path.

hwpang commented:
@davidegraff Thanks for the suggestions! I moved the notebook to the top level, and I will work on the CLI utility.

hwpang commented:
In the new commit I add the CLI for the model converter. I made both the `input_path` and `output_path` keyword arguments and set `input_path` to be required after discussing this with @kevingreenman.

KnathanM commented on chemprop/v2/cli/convert.py:
```diff
@@ -0,0 +1,55 @@
+def add_convert_args(parser: ArgumentParser) -> ArgumentParser:
+    parser.add_argument(
+        "-i",
+        "--input_path",
```
It could be good to keep to our v2 convention to use dashes for long args `--input-path`.

KnathanM commented on chemprop/v2/cli/convert.py:
```diff
@@ -0,0 +1,55 @@
+        "-o",
+        "--output_path",
+        type=str,
+        help="Path at which the converted model will be saved.",
```
A note about what the default is would be helpful in the help string. Something like "If not specified, defaults to the name of the input file suffixed with '_v2.ckpt' in your current working directory."

KnathanM commented on chemprop/v2/cli/convert.py:
```diff
@@ -0,0 +1,55 @@
+    return args
+
+def main(args: Namespace) -> None:
+    logging.info(f"Converting v1 model checkpoint '{args.input_path}' to v2 model checkpoint '{args.output_path}'...")
```
I believe to use logging you need this line after your imports at the top of the script `logger = logging.getLogger(__name__)` and then use `logger.info(...)` instead of `logging.info(...)`.

KnathanM commented on chemprop/v2/cli/convert.py:
```diff
@@ -0,0 +1,55 @@
+    parser = ArgumentParser()
+    ConvertSubcommand.add_args(parser)
+
+    logging.basicConfig(stream=sys.stdout, level=logging.DEBUG, force=True)
```
Same as above

KnathanM replied:
I was wrong about this one. It is `logging.basicConfig(...)`

hwpang commented:
@KnathanM Thanks for your comment! I made the changes as suggested.

davidegraff commented on chemprop/v2/cli/convert.py:
```diff
@@ -0,0 +1,58 @@
+        args = process_args(args)
+        main(args)
+
+def add_convert_args(parser: ArgumentParser) -> ArgumentParser:
+    parser.add_argument(
+        "-i",
+        "--input-path",
+        type=str,
+        required=True,
+        help="Path to a v1 model .pt checkpoint file.",
+    )
+    parser.add_argument(
+        "-o",
+        "--output-path",
+        type=str,
+        help="Path at which the converted model will be saved. If not specified, it defaults to the name of the input file suffixed with '_v2.ckpt' in your current working directory.",
+    )
+    return parser
```
can you just move this all into `ConvertSubcommand.add_args`?

davidegraff commented on chemprop/v2/cli/convert.py:
```diff
@@ -0,0 +1,58 @@
+        main(args)
+
+def add_convert_args(parser: ArgumentParser) -> ArgumentParser:
+    parser.add_argument(
+        "-i",
+        "--input-path",
+        type=str,
+        required=True,
+        help="Path to a v1 model .pt checkpoint file.",
+    )
```
```suggestion
    parser.add_argument(
        "input_path",
        help="Path to a v1 model .pt checkpoint file.",
    )
```
This makes `input-path` into a positional arg, which is always required. IMO it makes sense that a user calls this with only two arguments: (1) input path and (2, optional) output path

davidegraff commented on chemprop/v2/cli/convert.py:
```diff
@@ -0,0 +1,58 @@
+        help="Path to a v1 model .pt checkpoint file.",
+    )
+    parser.add_argument(
+        "-o",
+        "--output-path",
+        type=str,
```
```suggestion
        "output_path",
        nargs="?",
```
This makes `output_path` into an optional, positional argument (after `input_path`).

davidegraff commented on chemprop/v2/cli/convert.py:
```diff
@@ -0,0 +1,58 @@
+    )
+    return parser
+
+def process_args(args: Namespace) -> Namespace:
+    args.input_path = Path(args.input_path)
+    args.output_path = Path(args.output_path or Path.cwd() / (args.input_path.stem + "_v2.ckpt"))
+    return args
+
+def main(args: Namespace) -> None:
+    logging.info(f"Converting v1 model checkpoint '{args.input_path}' to v2 model checkpoint '{args.output_path}'...")
+    convert_model_v1_to_v2_file(str(args.input_path), str(args.output_path))
```
move all this code into `ConvertSubcommand.func`

davidegraff commented on chemprop/v2/utils/v1_to_v2.py:
```diff
@@ -0,0 +1,140 @@
+        "encoder.encoder.0.W_o.bias"
+    ]
+
+    state_dict_v2["readout.loc"] = torch.from_numpy(
+        model_v1["data_scaler"]["means"]
+    ).unsqueeze(0)
+    state_dict_v2["readout.scale"] = torch.from_numpy(
+        model_v1["data_scaler"]["stds"]
+    ).unsqueeze(0)
```
only `RegressionReadout` classes have the `loc` and `scale` attributes right?

davidegraff commented on chemprop/v2/utils/v1_to_v2.py:
```diff
@@ -0,0 +1,140 @@
+    args_v1 = model_v1["args"]
+    hyper_parameters_v2[
+        "batch_norm"
+    ] = False  # It doesn't seem that batch norm is done in v1
```
yes, batch norm is new in v2

KnathanM replied:
I think you could remove the comment or change it to " # Batch norm is new to v2". My preference is remove.
Also I don't know why this hyperparameter is spread over three lines. I think it could be collapsed to just one.
`hyper_parameters_v2["batch_norm"] = False`

hwpang replied:
Thanks! I remove the comment in the new push. I use `black` as my default formatter, and it tries to format the code if the code is too long to improve readability. In the new push I also make this into one line.

davidegraff commented on chemprop/v2/utils/v1_to_v2.py:
```diff
@@ -0,0 +1,140 @@
+    
+    model_v2 = {}
+    
+    model_v2["epoch"] = None  # Not defined in v1 model dictionary
+    model_v2["global_step"] = None  # Not defined in v1 model dictionary
+    model_v2["pytorch-lightning_version"] = __version__ 
+    model_v2["state_dict"] = convert_state_dict_v1_to_v2(model_v1)
+    model_v2["loops"] = None  # Not defined in v1 model dictionary
+    model_v2["callbacks"] = None  # Not defined in v1 model dictionary
+    model_v2["optimizer_states"] = None  # Not defined in v1 model dictionary
+    model_v2["lr_schedulers"] = None  # Not defined in v1 model dictionary
```
remove comments

KnathanM commented on chemprop/v2/cli/convert.py:
```diff
@@ -0,0 +1,58 @@
+    parser = ArgumentParser()
+    ConvertSubcommand.add_args(parser)
+
+    logger.basicConfig(stream=sys.stdout, level=logging.DEBUG, force=True)
```
Sorry Hao-Wei, I should have been more clear. The logging syntax is
`logging.basicConfig(...)`
`logger.info(...)`
These need to be changed on lines 47 and 54.

davidegraff commented on chemprop/v2/utils/v1_to_v2.py:
```diff
@@ -0,0 +1,140 @@
+def convert_model_v1_to_v2_file(model_v1_file: str, model_v2_file: str) -> None:
+    """Converts a v1 model .pt file to a v2 model .ckpt file
+    """
+    model_v1 = torch.load(model_v1_file)
+    model_v2 = convert_model_v1_to_v2(model_v1)
```
can we make the variable names more clear? `model_v*` makes it sound like these variables are actual models/DMPNNs, but these are just dictionaries containing model (hyper)parameters?

davidegraff commented on chemprop/v2/cli/convert.py:
```diff
@@ -0,0 +1,58 @@
+
+def process_args(args: Namespace) -> Namespace:
+    args.input_path = Path(args.input_path)
+    args.output_path = Path(args.output_path or Path.cwd() / (args.input_path.stem + "_v2.ckpt"))
```
I think we want the default output path to be like so: `input_path = path/to/model.pt` → `output_path = path/to/model_v2.ckpt`. Right now it will write to `$CWD/model_v2.ckpt`

davidegraff reviewed (commented):
a bunch of tiny comments, but happy to approve after they're addressed

KnathanM commented on chemprop/v2/cli/convert.py:
```diff
@@ -0,0 +1,44 @@
+        parser.add_argument(
+            "output_path",
+            nargs="?",
+            help="Path at which the converted model will be saved. If not specified, it defaults to the name of the input file suffixed with '_v2.ckpt' in your current working directory.",
```
We changed the default:
"If not specified, it defaults to the name of the input file suffixed with '_v2.ckpt' in the same parent directory."

I think how I worded this earlier could be confusing. Maybe a simple example is more clear:
"If not specified and the input file is `/path/to/checkpoint/checkpoint.pt`, the output will default to `/path/to/checkpoint/checkpoint_v2.ckpt`. 

Please choose what sounds clearer to you.

KnathanM commented on chemprop/v2/cli/convert.py:
```diff
@@ -0,0 +1,44 @@
+        args.output_path = Path(args.output_path or args.input_path.parent / (args.input_path.stem + "_v2.ckpt"))
+
+        logger.info(f"Converting v1 model checkpoint '{args.input_path}' to v2 model checkpoint '{args.output_path}'...")
+        convert_model_file_v1_to_v2(str(args.input_path), str(args.output_path))
```
`args.input_path` starts as a string. You then convert it to a path using `Path()`. The on line 35 you convert it back to a string. Would it be okay to just leave `args.input_path` and `args,output_path` as strings?

davidegraff replied:
can the converter function not accept `Path`s?

hwpang replied:
I thought it couldn't, but I just tested and it seemed to work fine with `Paths`, so I changed this part of code to pass just `args.input_path` and `args.output_path` in without converting them into `str`.

hwpang commented:
@KnathanM @davidegraff Thank you so much for the comments! In the most recent push, I have addressed all the comments and cleaned up the commits. I also add a unit test for the converter to test that we can load the converted model.

hwpang commented:
@KnathanM I just saw your newer comments. I will work on them now.

hwpang commented:
@KnathanM I address your comments in the new push! I think this PR is ready for review and merge.

davidegraff commented on chemprop/v2/cli/convert.py:
```diff
@@ -0,0 +1,44 @@
+        parser.add_argument(
+            "output_path",
+            nargs="?",
+            help="Path at which the converted model will be saved. If not specified and the input file is /path/to/checkpoint/model.pt, the output will default to /path/to/checkpoint/model_v2.ckpt",
```
```suggestion
            help="The path to which the converted model will be saved. If not specified and the input file is '/path/to/checkpoint/model.pt', the output will default to '/path/to/checkpoint/model_v2.ckpt'",
```

davidegraff commented on chemprop/v2/tests/unit/utils/test_converter.py:
```diff
@@ -0,0 +1,20 @@
+def example_model_v1_path():
+    return TEST_DATA_DIR / "example_model_v1.pt"
+
+def test_converter(example_model_v1_path):
+    convert_model_file_v1_to_v2(example_model_v1_path, "example_model_v2.ckpt")
```
use the pytest `tmp_path` [fixture](https://docs.pytest.org/en/6.2.x/tmpdir.html) here so that you don't actually need to clean up the written checkpoint. The reason this matters is that this test, as written, could fail for file writing/removal, which is outside the scope of the test.

davidegraff reviewed (commented):
Broad question, are there other ways we can test this utility? I can't think of any right now, but simply checking that we indeed write a file seems a little light. No worries if not and happy to accept after the changes, but something to keep in mind.

davidegraff commented:
Also, you're just running `black` with the default settings right (i.e., not supply any additional flags)? This ensures that you use the default settings from `pyproject.toml`

hwpang commented:
@davidegraff Thanks for the comments! In the new push, I (1) update the unit test to test that the converted model can make predictions sufficiently close to the original model and (2) use `tmp_path` to create a temporary path for the checkpoint file made by the converter.

davidegraff reviewed (approved):
👍  good work, thanks!
